src/main/kotlin/io/bazel/kotlin/builder/tasks/jvm/CompilationTask.kt

```diff
@@ -161,191 +198,116 @@ internal fun JvmCompilationTask.kaptArgs(
+            // It will be added by annotationProcessingArgs() below
+            classpath = inputs.stubsPluginClasspathList.filterNot { it == pluginJar },
+          ),
+        ).plus(
+          annotationProcessingArgs(context, config, pluginJar),
+        ).flag("-d", directories.generatedClasses)
+        .values(inputs.kotlinSourcesList)
+        .values(inputs.javaSourcesList)
+        .list()
+
+    args
+      .let { compilationArgs ->
         context.executeCompilerTask(
-          args,
+          compilationArgs,
           compiler::compile,
           printOnSuccess = context.whenTracing { true } == true,
         )
       }.let { outputLines ->
-        // if tracing is enabled the output should be formatted in a special way, if we aren't
-        // tracing then any compiler output would make it's way to the console as is.
         context.whenTracing {
-          printLines("kapt output", outputLines)
+          context.printCompilerOutput(listOf("${config.pluginId} output:") + outputLines)
         }
         return@let expandWithGeneratedSources()
       }
   }
 }
 
-private fun JvmCompilationTask.runKspPlugin(
+internal fun JvmCompilationTask.runPlugins(
   context: CompilationTaskContext,
   plugins: InternalCompilerPlugins,
   compiler: KotlinToolchain.KotlincInvoker,
 ): JvmCompilationTask {
-  return context.execute("Ksp (${inputs.processorsList.joinToString(", ")})") {
-    val overrides =
-      mutableMapOf(
-        API_VERSION_ARG to kspKotlinToolchainVersion(info.toolchainInfo.common.apiVersion),
-        LANGUAGE_VERSION_ARG to
-          kspKotlinToolchainVersion(
-            info.toolchainInfo.common.languageVersion,
-          ),
-      )
-    baseArgs(overrides)
-      .plus(kspArgs(plugins))
-      .flag("-d", directories.generatedClasses)
-      .values(inputs.kotlinSourcesList)
-      .values(inputs.javaSourcesList)
-      .list()
-      .let { args ->
-        context.executeCompilerTask(
-          args,
-          compiler::compile,
-          printOnSuccess = context.whenTracing { true } ?: false,
-        )
-      }.let { outputLines ->
-        // if tracing is enabled the output should be formatted in a special way, if we aren't
-        // tracing then any compiler output would make it's way to the console as is.
-        context.whenTracing {
-          printLines("Ksp output", outputLines)
-        }
-        return@let expandWithGeneratedSources()
-      }
+  // Early exit if no annotation processing or sources
+  if (!inputs.hasAnnotationProcessing() || inputs.kotlinSourcesList.isEmpty()) {
+    return this
   }
-}
 
-private fun kspKotlinToolchainVersion(version: String): String {
-  // KSP doesn't support Kotlin 2.0 yet, so we need to use 1.9
-  return if (version.toFloat() >= 2.0) "1.9" else version
+  // Generic annotation processing path (KAPT, etc.)
+  val config = inputs.annotationProcessing
+
+  // Extract plugin jar from classpath (provided by Starlark layer)
```

Comment:
It would be better to either pass the plugin jar as a separate argument, or generate the -Xplugin flag in starlark.